src/engine.cpp

```diff
@@ -874,6 +875,45 @@ static void DisableEngineForCompany(EngineID eid, CompanyID company)
 	}
 }
 
+/** Vanilla implementation of Engine::WasPreviewSuccessful. */
+static bool VanillaWasPreviewSuccessful(CompanyID c_index, EngineID e_index) {
+	/* Check the company's 'ALL_GROUP' group statistics. This only includes countable vehicles, which is fine
+	 * as those are the only engines that can be given exclusive previews. */
+	return GetGroupNumEngines(c_index, ALL_GROUP, e_index) != 0;
+}
+
+/** Vanilla implementation of Engine::OnPreviewFailed. */
+static void VanillaOnPreviewFailed(Company *c, Engine *e, EnginePreviewNewsInformation *epni) {
+	c->block_preview = 20;
+
+	/* Give higher fine to bigger companies. */
+	Money fine = static_cast<Money>(pow((GetGroupNumVehicle(c->index, ALL_GROUP, e->type) + 7)/16.0, 2) * static_cast<double>(e->GetCost()));
```

Comment:
Setting the aside the question of doing this at all, it is not valid to use non-deterministic float arithmetic in this context. The calculated cost needs to be deterministic and identical on the server and all clients.